sklearn/linear_model/_sgd_fast.pyx.tp

```diff
@@ -1,22 +1,50 @@
-# Author: Peter Prettenhofer <peter.prettenhofer@gmail.com>
-#         Mathieu Blondel (partial_fit support)
-#         Rob Zinkov (passive-aggressive)
-#         Lars Buitinck
-#
-# License: BSD 3 clause
+{{py:
 
+"""
+Template file to easily generate fused types consistent code using Tempita
+(https://github.com/cython/cython/blob/master/Cython/Tempita/_tempita.py).
 
+Generated file: _sgd_fast.pyx
+
+Each relevant function is duplicated for the dtypes float and double.
+The keywords between double braces are substituted in setup.py.
+
+Authors: Peter Prettenhofer <peter.prettenhofer@gmail.com>
+         Mathieu Blondel (partial_fit support)
+         Rob Zinkov (passive-aggressive)
+         Lars Buitinck
+
+License: BSD 3 clause
+"""
+
+# The dtypes are defined as follows (name_suffix, c_type, np_type)
+dtypes = [
+    ("64", "double", "np.float64"),
+    ("32", "float", "np.float32"),
+]
+
+}}
+
+#------------------------------------------------------------------------------
+
+"""
+SGD implementation
+WARNING: Do not edit .pyx file directly, it is generated from .pyx.tp
+"""
+
+from cython cimport floating
 import numpy as np
 from time import time
 
 from libc.math cimport exp, log, pow, fabs
 cimport numpy as cnp
 from numpy.math cimport INFINITY
 cdef extern from "_sgd_fast_helpers.h":
-    bint skl_isfinite(double) nogil
+    bint skl_isfinite32(float) nogil
+    bint skl_isfinite64(double) nogil
 
-from ..utils._weight_vector cimport WeightVector64 as WeightVector
-from ..utils._seq_dataset cimport SequentialDataset64 as SequentialDataset
+from ..utils._weight_vector cimport WeightVector32, WeightVector64
```

Comment:
When I put on my "numerical computing on a CPU hat", I tend to think that `float32` input dtypes should have accumulators of `float64`. Concretely, `WeightVector32` keeps internal statistics as `float64` and upcast data as it is processing the `float32` input. In any case, I think this PR can be accepted as is and any updates to `WeightVector32` can be done as a follow up.

@jjerphan @jeremiedbb What do you think?

XREF: For reference, PyTorch uses a `float64` accumulator for `float32` inputs on [CPUS](https://github.com/pytorch/pytorch/blob/b89fda51cd3899eae1dc9e5a2f2700c4113a4400/aten/src/ATen/AccumulateType.h#L137-L140) and `float32` accumulator on `float32` inputs for [GPUs](https://github.com/pytorch/pytorch/blob/b89fda51cd3899eae1dc9e5a2f2700c4113a4400/aten/src/ATen/AccumulateType.h#L69-L72)

Reply:
I opened https://github.com/scikit-learn/scikit-learn/pull/25721 to force `WeightVector32` to use a `float64` accumulator.